index.js

```diff
@@ -1,47 +1,246 @@
-    throw new Error('hyperdiscovery swarm must be created after the local hyperdb instance is ready!')
-  }
-
-  // Discovery Swarm Options
-  opts = xtend({
-    port: 3282,
-    id: isHyperdbInstance ? archive.local.id.toString('hex') : archive.id,
-    hash: false,
-    stream: function (peer) {
-      return archive.replicate(xtend({
-        live: self.live,
-        upload: self.uploading,
-        download: self.downloading
-      }, isHyperdbInstance ? {
-        userData: archive.local.key
-      } : {}))
-    }
-  }, opts)
-
-  this.swarm = disc(swarmDefaults(opts))
-  this.swarm.once('error', function () {
-    self.swarm.listen(0)
-  })
-
-  this.swarm.listen(opts.port)
-  this.swarm.join(this.archive.discoveryKey)
-  return this.swarm
+const crypto = require('crypto')
+const EventEmitter = require('events')
+
+const datEncoding = require('dat-encoding')
+const hypercoreProtocol = require('hypercore-protocol')
+const discoverySwarm = require('discovery-swarm')
+const swarmDefaults = require('dat-swarm-defaults')
+const mutexify = require('mutexify')
+const getPort = require('get-port')
+
+const debug = require('debug')('hyperdiscovery')
+
+module.exports = (...args) => new Hyperdiscovery(...args)
+
+const DEFAULT_PORT = 3282
+const ALTERNATE_PORTS = [3000, 3002, 3004, 2001, 2003, 2005]
+
+class Hyperdiscovery extends EventEmitter {
+  // Modified from Beaker Browser, copyright Blue Link Labs:
+  //    https://github.com/beakerbrowser/beaker-core/
+  //    https://github.com/beakerbrowser/dat-node
+  // And Core Store, copyright Andrew Osheroff:
+  //    https://github.com/andrewosh/corestore
+
+  constructor (feed, opts) {
+    super()
+
+    if (feed && !feed.replicate) {
+      opts = feed
+      feed = null
+    }
+    opts = opts || {}
+    // Old Options:
+    // * `stream`: function, replication stream for connection. Default is `archive.replicate({live, upload, download})`.
+    // * `upload`: bool, upload data to the other peer?
+    // * `download`: bool, download data from the other peer?
+    // * `port`: port for discovery swarm
+    // * `utp`: use utp in discovery swarm
+    // * `tcp`: use tcp in discovery swarm
+
+    this._opts = opts
+    this.id = opts.id || crypto.randomBytes(32)
+    this.port = typeof opts.port === 'number' ? opts.port : DEFAULT_PORT
+
+    this._swarm = discoverySwarm(swarmDefaults({
+      id: this.id,
+      hash: false,
+      utp: defaultTrue(opts.utp),
+      tcp: defaultTrue(opts.tcp),
+      dht: defaultTrue(opts.dht),
+      stream: this._createReplicationStream.bind(this)
+    }))
+
+    // bubble listening and errors
+    this._swarm.on('listening', () => {
+      this.port = this._swarm.address().port
+      this.emit('listening', this.port)
+      debug('swarm:listening', { port: this.port })
+    })
+    this._swarm.on('error', async (err) => {
+      if (err && err.code !== 'EADDRINUSE') return this.emit('error', err)
+      const port = await getPort({ port: ALTERNATE_PORTS })
```

Comment:
getPort actually creates a server and listens on the port. I do not recommend this module for that usecase. as mentioned in https://github.com/sindresorhus/get-port#beware

Reply:
Also: `ALTERNATE_PORTS` should be a option.

Reply:
Yea, I was aware it created a server since I think thats the only way to check if a port is open (please correct me if wrong). So we will just create that server ourself =)